### PR TITLE
[SYCL][DOC] Introduce sycl-marray-complex specialization proposal

### PR DESCRIPTION
**Pull request by jle-quel**
Following the proposal for `sycl::ext::oneapi::experimental::complex` and its implementation (https://github.com/intel/llvm/pull/8068), which adds the support of complex for SYCL, this proposal allows for adding new complex math features.

This PR proposes the specialization of `sycl::marray` to add support for storing complex numbers in arrays.

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  operator ComplexDataT() const;
+
+  static constexpr std::size_t size() noexcept;
+
+  // real and imag
+  marray<T, NumElements> real() const;
+  marray<T, NumElements> imag() const;
+
+  // subscript operator
+  reference operator[](std::size_t i);
```
Comment (rolandschulz) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:153]: This requires that the internal data representation is an array of complex. Are we reasonable sure no implementation wants the data storage to be one array for real and one for imag?

Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:153]: Please, let me know if I misunderstood your feedback, but that's what the sycl::marray specialization for sycl::complex is about.
If needed, a user could use the `real()` or `imag()` method to let him work on the data split.

@TApplencourt, what do you think?

Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:153]: There is already an implementation of this complex specialization of marray within this repo https://github.com/argonne-lcf/SyclCPLX.
No issues have been raised about having an array of complex rather than two arrays for `real` and `imag`.

Also, this PR is for the proposed directory and later will be moved onto the experimental.
We'll not be tied to this proposal implementation.

I believe it is conceivable to have an array of complex for this proposal as a start, and if needed we'll change the structure layout.

**Comment by jle-quel** [comment]
Gentle ping to get some attention into this PR :+1: @rolandschulz @steffenlarsen @gmlueck

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+`sycl::marray` class defines is only implemented if the marray's value type also
+implements the operator.
+
+For example, `sycl::marray<sycl::ext::oneapi::complex, N>` does not implement
+the modulus operator as `sycl::ext::oneapi::complex` does not support it.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:116]: The [complex extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc) is experimental, so this should be in the experimental namespace too as it relies on it.

**Comment by gmlueck** [comment]
Ack!  Sorry, I just noticed that my previous comments never got submitted.  (They were stuck as "pending" in the PR.)  Sorry about the delay.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 2 above)
Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:116]: The templated class `marray` has been declared and defined under the namespace `sycl/ext/oneapi`.

Correct me if I'm wrong, but an explicit specialization can only be done in the same namespace scope as the primary template.

Comment (steffenlarsen) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:116]: Ah, I see what you mean. I suppose this falls into the category of extension behavior of existing interfaces. Tag @gmlueck & @rolandschulz .

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+Finally, each math function between each element should follow the C++
+standard for handling `NaN` and `Inf` values.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+// Compute the magnitude for each complex number in marray x.
+mgenfloat abs(const mgencomplex& x);
```
[the remark this answers is not in the document]
Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:283]: The reason that this function is called `abs` is due to the heritage of `std::complex`.

`abs` is one of `std::complex`'s non-member functions.
The  `sycl::ext::oneapi::experimental::complex` has this equivalent non-member function.

The math `marray` overloads these functions and thus inherits the name.

The question is whether we would like to change `abs` to `fabs` for `complex`, but `experimental::complex` being a port of `std::complex`, and an easy replacement.
Modifying the interface of our `complex` will add errors when switching from `std::complex` to `experimental::complex`.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+|`mgenfloat imag() const;`
+|Returns a marray of the imaginary components for marray of complex numbers held by this `marray`.
+|===
+
+=== Mathematical operations
+
+This proposal extends `sycl::ext::oneapi` namespace math functions to accept
+`mgencomplex` for the SYCL math functions, `abs`, `acos`, `asin`, `atan`,
+`acosh`, `asinh`, `atanh`, `arg`, `conj`, `cos`, `cosh`, `exp`, `log`, `log10`,
+`norm`, `polar`, `pow`, `proj`, `sin`, `sinh`, `sqrt`, `tan`, and `tanh`.
```
[the remark this answers is not in the document]
Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:262]: These functions are the overloads of the complex's math functions, proposed [there](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc).

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+  friend marray<ComplexDataT, NumElements> operator~(const marray<ComplexDataT, NumElements> &v) = delete;
+
+  friend marray<bool, NumElements> operator!(const marray<ComplexDataT, NumElements> &v) = delete;
+};
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:234]: Actually, I wonder why we want to add new member functions to `marray` at all.  Rather than doing this, let's add two new free functions like:

```
template<typename T, size_t NumElements>
marray<T, NumElements> real(const marray<complex<T>, NumElements> &z);

template<typename T, size_t NumElements>
marray<T, NumElements> imag(const marray<complex<T>, NumElements> &z);
```

This is consistent with standard C++, which has free functions `std::real(const std::complex<T> &z)` and `std::imag(const std::complex<T> &z)`.  There is no "marray" in standard C++, so there is no precedent for extending it with extra member functions.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:283]: I see.  Makes sense to follow C++ here.

**Comment by gmlueck** [comment]
One other thought ... Why should this be a separate extension rather than an expansion of the existing [sycl_ext_oneapi_complex](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc)?

Of course you can argue that more granular extensions give implementors greater freedom to choose what they implement, but more extensions makes life harder for application developers who want to write portable code.

Consider an implementor who already implements sycl_ext_oneapi_complex.  Is there some reason that implementor might not be able to implement these marray functions?  It seems like it's just a matter of some extra code on top of what they need already for sycl_ext_oneapi_complex.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:234]: I understand your point and here's a little bit of context.

The reason for the `real` and `imag` being members of the `marray` specialization is the conclusion of a discussion that happened within this PR https://github.com/intel/llvm/pull/6550.

The goal was to have equivalent members within `marray`, so this struct could be abstract and you will apply the same member on the value that you are working on as if this was a `complex`.

Because `complex` has a `real` and `imag` member, `marray` has a `real` and `imag` member.

**Comment by jle-quel** [comment]
> One other thought ... Why should this be a separate extension rather than an expansion of the existing [sycl_ext_oneapi_complex](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc)?
> 
> Of course you can argue that more granular extensions give implementors greater freedom to choose what they implement, but more extensions makes life harder for application developers who want to write portable code.
> 
> Consider an implementor who already implements sycl_ext_oneapi_complex. Is there some reason that implementor might not be able to implement these marray functions? It seems like it's just a matter of some extra code on top of what they need already for sycl_ext_oneapi_complex.

You are right, the best definition of that proposal is an expansion of the existing complex. Bringing other types which handle complex and overloads for the complex's math functions to support this new type.

No, there is no reason that an implementor might not be able to implement these `marray` functions, considering that he has an `marray` implementation.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:234]: There's a lot of discussion in #6550.  Can you summarize the reason for adding these member functions to `marray`?  Much of the discussion I see there is arguing to keep the API simple in this extension.  It seems to me that the API is simpler if we do not add new member functions to `marray`.

> The goal was to have equivalent members within `marray`, so this struct could be abstract and you will apply the same member on the value that you are working on as if this was a `complex`.

I'm not quite sure what you are arguing here.  Do you want to allow generic code that operates on either an `marray` of `complex` or on a single `complex`?  I'd recommend adding free functions `real` and `imag` functions that operate on either like:

```
template<typename T>
T real(const complex<T> &z);

template<typename T>
T imag(const complex<T> &z);

template<typename T, size_t NumElements>
marray<T, NumElements> real(const marray<complex<T>, NumElements> &z);

template<typename T, size_t NumElements>
marray<T, NumElements> imag(const marray<complex<T>, NumElements> &z);
```

This way, the generic code can use the free functions, not the member functions.

[the remark this answers is not in the document]
Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:234]: That was the goal yes "allow generic code that operates on either an marray of complex or on a single complex".

Alright, So this would be the real/imag API:

```
complex.real();
complex.imag();
real(complex)
imag(complex);

// newest member real and imag are being deleted from the marray specialization

real(marray<complex>);
imag(marray<complex>);
```

[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:234]: right!

**Comment by steffenlarsen** [comment]
@gmlueck & @jle-quel - What are the next steps here?

**Comment by jle-quel** [comment]
> @gmlueck & @jle-quel - What are the next steps here?

On my side, I got back to work on this today.
I'll update the `"getter" real/imag API` as discussed with Greg, as well as the other texts/typos feedback.

Regarding the implementation of this proposal (https://github.com/intel/llvm/pull/8647) I got back to it today as well, and we'll update the `getter` API, and also split the single header files into multiples.

The next steps I'm working towards and hoping for would be:
- The proposal would be up to date with the latest feedback.
- I am still unsure about the separate extension versus expansion (Don't have a strong opinion on it).
- The implementations would be up to date as well with the implementation.
- Finally, a final review before we might merge the `marray's complex` specialization (doc and impl)

**Comment by jle-quel** [comment]
Since we are moving away from having a separate extension, I'll create another PR to expand the complex extension with the modifications asked here.
I'll link the two together for the future reader to know where the discussion went, and will close this PR.
